src/Browscap/Filter/Property/FullPropertyFilter.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Browscap\Filter\Property;
+
+use Browscap\Data\PropertyHolder;
+use Browscap\Writer\WriterInterface;
+
+/**
+ * this filter is responsible to select properties for the "full" version of the browscap files
+ */
+trait FullPropertyFilter
+{
+    /** @throws void */
+    public function __construct(private PropertyHolder $propertyHolder)
```

Comment:
I dont like the idea of having a constructor inside a trait.

Reply:
I did consider whether it was a good idea, but a few things made me decide to put it here:

- The self contained nature of the trait, and code deduplication. Without this, everywhere using this trait would have to setup the `propertyHolder` class property, as the required `PropertyHolder` type. With this it just works by being used in classes.
- With the self contained nature, it also makes the trait functional and testable. The trait shouldn't need to rely on structure in classes using it to be functional / testable, only the defined dependency injection required by the constructor.
- The empty nature of the PHP8 constructor, with it's argument. This means there's no code to duplicate / no reason to try to call `parent::__construct` (a primary reason not to include a constructor in a trait). If engineers want to overwrite the constructor, it's a pretty simple process to replicate the functionality anyway, which they would with simply adding an argument of `private PropertyHolder $propertyHolder`. If it were a parent class instead, they'd need a constructor argument of `PropertyHolder $propertyHolder` on the child, and this would need to be passed to the `parent::__construct` call.

Reply:
Hmm, I need to look into this in more detail, but don't currently have the time; in general, introducing any `trait` at all is likely an antipattern. I need to review properly when I have some time to see exactly what you're trying to do here, and whether a better solution can be in place.

Reply:
Full disclosure, I did for-see a better solution, but it breaks BC. These traits would be classes in their own right instead, and the original filter classes where the trait versions are used should be taken away. I say this down to the single responsibility principle, and as mentioned above, there not being any proper link between Divisions / Sections, & Properties.

As there's noting in the v7 milestone, I was trying to avoid breaking BC, and find a half way house between the existing code, and the ultimate destination of completely separate classes. I saw that as a way of getting this functionality in sooner, but equally honestly, should raised a discussion on the subject instead, to see what the requirements are for a new major version.

Reply:
This is very similar to the `CustomFieldStandardFilter` class in my original PR description, where I was trying to use to demonstrate what people required to get round this current bug. It feels very much like an anti-pattern itself, injecting one filter into another, to try and sort a bug, and one I was trying to avoid when creating the traits. I'd have thought filters should be combined via a `FilterCollection` class instead, but then that's what helped to identify this bug. 🤷‍♂️